### ISSUE TITLE
fails locate hsdev backend - Ubuntu 16.04 with stack haskell

### ISSUE THREAD

**Issue by merflogin** [SublimeHaskell/SublimeHaskell#376]
SublimeHaskell suddenly stopped working - it looked like it wasn't able to find backend (I used `hsdev`), 
Here are steps that I've tried to make him work again. They didn't help. Any suggestions?

I've removed plugin through Package Control. Closed Sublime. And installed plugin back again.
I've received the response:

> No usable backends (hsdev, ghc-mod) found on PATH or
> in the standard Haskell cabal and stack installation
> locations.
> 
> Please check or update your SublimeHaskell user settings,
> or install hsdev or ghc-mod.

while `whereis` clearly indicates that `hsdev` is still where it were
```
merf@merf-GL552VW:~$ whereis hsdev
hsdev: /home/merf/.local/bin/hsdev
merf@merf-GL552VW:~$ hsdev version
0.2.5.1
```

While it is a standard stack installation place and doesn't need to be added to PATH I've put it in user settings anyway just to be sure:
```
{
	"add_to_PATH": ["/home/merf/.local/bin"]
}
```
Restart of sublime showed that `zzplugin` fails to load while trying to get version of `hsev` - it can't get `install_dir`.
Here is sublime console
 ```
...
reloading plugin SublimeHaskell.types
reloading plugin SublimeHaskell.zzplugin
plugins loaded
Traceback (most recent call last):
  File "/opt/sublime_text/sublime_plugin.py", line 210, in on_api_ready
    m.plugin_loaded()
  File "/home/merf/.config/sublime-text-3/Packages/SublimeHaskell/zzplugin.py", line 34, in plugin_loaded
    backend_mgr.get_backends()
  File "/home/merf/.config/sublime-text-3/Packages/SublimeHaskell/internals/backend_mgr.py", line 140, in get_backends
    self.possible_backends = self.filter_possible(Settings.PLUGIN.backends)
  File "/home/merf/.config/sublime-text-3/Packages/SublimeHaskell/internals/backend_mgr.py", line 208, in filter_possible
    if ubclazz is not None and ubclazz.is_available(**ubackend.get('options', {})):
  File "/home/merf/.config/sublime-text-3/Packages/SublimeHaskell/hsdev/backend.py", line 93, in is_available
    hsdev_ver = HsDevBackend.hsdev_version(kwargs.get('exec-with'), kwargs.get('install-dir'))
  File "/home/merf/.config/sublime-text-3/Packages/SublimeHaskell/hsdev/backend.py", line 212, in hsdev_version
    hsdev_proc = ProcHelper.exec_with_wrapper(exec_with, install_dir, ['hsdev', 'version'])
  File "/home/merf/.config/sublime-text-3/Packages/SublimeHaskell/internals/proc_helper.py", line 175, in exec_with_wrapper
    raise RuntimeError('ProcHelper.exec_with_wrapper: invalid install_dir (None)')
RuntimeError: ProcHelper.exec_with_wrapper: invalid install_dir (None)
```

**Comment by bscottm** [comment]
@merflogin: What does your user configuration for `backends` look like? `invalid install_dir` tends to indicate that you're trying to use a `stack` or `cabal`-wrapped `hsdev`, but haven't set all of the options.

**Comment by merflogin** [comment]
@bscottm, I haven't changed `backends` dictionary if that's what you're asking. It's not in the user settings so it should (?) use the default "Local hsdev" (bellow).

It's not done intentionally, but I assumed that plugin would use `hsdev` executable without `stack` wrapper.

```
"backends": {
        "Local hsdev": {
            "default": true,
            "backend": "hsdev",
            "options": {
              "local": true,
              "port": 4567
            }
        },
        "ghc-mod": {
            "default": false,
            "backend": "ghc-mod"
        },
        "Remote hsdev@localhost:4567": {
            "default": false,
            "backend": "hsdev",
            "options": {
                "local": false,
                "host": "localhost",
                "port": 4567
            }
        },
        "Cabal-sandboxed local hsdev": {
            "default": false,
            "backend": "hsdev",
            "options": {
              "local": true,
              "port": 4567,
              "exec-with": "cabal"
              // ^^^^^^^^^^^^^^^^^ Remember to insert a comma (",") after this line once you set install-dir
              // to hsdev's installation directory.
              //
              // Force a diagnostic message because this has to be set in conjunction with exec-with:
              // "install-dir": "$HOME/totally/invalid/directory"
            }
        },
        "Stack-exec local hsdev": {
            "default": false,
            "backend": "hsdev",
            "options": {
              "local": true,
              "port": 4567,
              "exec-with": "stack"
              // ^^^^^^^^^^^^^^^^^ Remember to insert a comma (",") after this line once you set install-dir
              // to hsdev's installation directory.
              //
              // Force a diagnostic message because this has to be set in conjunction with exec-with:
              // "install-dir": "$HOME/totally/invalid/directory"
            }
        },
        "Cabal new-build local hsdev": {
            "default": false,
            "backend": "hsdev",
            "options": {
              "local": true,
              "port": 4567,
              "exec-with": "cabal-new-build"
              // ^^^^^^^^^^^^^^^^^ Remember to insert a comma (",") after this line once you set install-dir
              // to hsdev's installation directory.
              //
              // Force a diagnostic message because this has to be set in conjunction with exec-with:
              // "install-dir": "$HOME/totally/invalid/directory"
            }
        },
        "No backend": {
            "default": false,
            "backend": "none"
        }
    }
```

**Comment by bscottm** [comment]
@merflogin: Duplicate of #375. Not sure what the particular issue is -- the workaround in the meantime is adding an `install_dir` setting to the "Local hsdev" backend dictionary. It'll take me a bit to investigate and find the bug.

**Comment by merflogin** [comment]
@bscottm, I thought that some additional input might help.

- What worked?
My workaround, for now, is to comment all backends except "Local hsdev". It removes error from the console and all features that I've spoted (autocompleteion, lint, etc.) is working.
```
reloading plugin SublimeHaskell.zzplugin
plugins loaded
plugin 'backends' ['Local hsdev']
Possible/usable 'backends': ['Local hsdev']
```
Menu entry in tools is still broken as in [#357](https://github.com/SublimeHaskell/SublimeHaskell/issues/375) though.

- What didn't worked?
The workaround from [#357](https://github.com/SublimeHaskell/SublimeHaskell/issues/375) didn't help. Changing `backends` to below resulted in the same error as before. 

```
"backends": {
        "Local hsdev": {
            "default": true,
            "backend": "hsdev",
            "options": {
              "local": true,
              "port": 4567,
              "install-dir": "/home/merf/.local/bin"
            }
        },
...
       "Stack-exec local hsdev": {
            "default": false,
            "backend": "hsdev",
            "options": {
              "local": true,
              "port": 4567,
              "exec-with": "stack",
              // ^^^^^^^^^^^^^^^^^ Remember to insert a comma (",") after this line once you set install-dir
              // to hsdev's installation directory.
              //
              // Force a diagnostic message because this has to be set in conjunction with exec-with:
              // "install-dir": "$HOME/totally/invalid/directory"
              "install-dir": "/home/merf/.local/bin"
            }
        },
...
    }
```

**Comment by bscottm** [comment]
@merflogin: What ST3 version are you using? I've tested with ST3, Build 3143 and Build 3156 (dev channel), can't seem to reproduce this issue. Will continue to investigate.

**Comment by bscottm** [comment]
@merflogin, @paxter: Can you put this into your SublimeHaskell user preferences, restart ST3/SublimeHaskell and post what you get in the console window? (@merflogin: Back out all of your workarounds, if it's not too painful so I can trace where things are going/have gone.)

    "log": 5,

Thanks. I think I may have found something in the `haden` backend's client, but can't really be sure unless I have something to which I can compare it.

**Comment by merflogin** [comment]
@bscottm, ST Version 3.0, Build 3143.

Now it's get weird. It even shows that it found hsdev and recognized its version in console with log 5

Sublime sttings:
```
{
    // Enable auto completion of module imports
    "auto_complete_imports": true,

    // Enable auto completion of LANGUAGE pragmas
    "auto_complete_language_pragmas": true,

    // Enable auto build project on save
    "enable_auto_build": false,

    // Haskell build system. Choices are:
    //
    // "cabal": Use cabal to build projects
    // "cabal-new-build": Use the new cabal build system
    // "stack" Use stack to build projects
    //
    // Ideally, you should put your default preference in your User preferences.
    //
    // May be set on a per-project basis in SublimeHaskell project settings.
    "haskell_build_tool": "stack",

    // Enable auto check on save
    "enable_auto_check": true,

    // Enable auto lint on save (as hints when building)
    "enable_auto_lint": true,

    // How to build the cabal project in order to obtain error messages.
    // This option exists because `cabal build` is slow and cannot re-print
    // warnings of already compiled modules.
    //
    // (Also see https://github.com/haskell/cabal/issues/1179, cabal 1.16).
    //
    // Possible values:
    //
    // - "normal": A normal build using the haskell_build_tool builder. This is
    //   a full build of the project.
    //
    // - "normal-then-warnings": A normal build that also collects project-wide
    //   warnings. This runs the builder tool twice.
    //
    // - "normal-then-test": A normal build that runs tests after a successful
    //   build.
    //
    // - "normal-then-bench": A normal build that runs benchmarks after a
    //   successful build.
    //
    // - "typecheck": Performs a type check of the whole project, trying to not
    //   do any more than necesssary to give faster feedback. Consider turning
    //   auto-build off and enable auto-check instead.
    //
    // - "typecheck-then-warnings": Like "typecheck", but afterwards collects
    //    warnings like "normal-then-warnings".
    //
    // Please note that it is currently impossible to properly use a `-fno-code`
    // build on a cabal project that links a shared library; it errors too early.
    // (https://github.com/haskell/cabal/issues/1176, cabal 1.16).
    //
    // This can be a project-specific setting (see per-project settings at the end
    // of this document.)
    "auto_build_mode": "normal-then-warnings",

    // Use the improved/alternate Haskell syntax?
    //
    // SublimeHaskell ships with an improved/alternate syntax highlighter. If this setting
    // 'true', Haskell files will use the new highlighter. If false, Haskell files will
    // revert to the Haskell highlighter shipped with ST.
    "use_improved_syntax": true,

    // Show output window on clean/configure/build/install:
    "show_output_window": true,

    // Show error window on build/check/lint:
    "show_error_window": true,

    // Add the "standard" directories where cabal and stack install executables,
    // if they exist, in the following order (top to bottom):
    //
    // stack's install dir
    // user bindir from $HOME/.cabal/config (*nix) or  %APPDATA%/cabal/config (WinXX)
    // global bindir from $HOME/.cabal/config (*nix) or %APPDATA%/cabal/config (WinXX)
    //
    // The defaults are:
    //
    // stack:  {$HOME/.local|%APPDATA%/local}/bin
    // user:   {$HOME/.cabal|%APPDATA%/cabal}/bin
    // global: {/usr/local|%PROGRAMFILES%/Haskell}/bin
    "add_standard_dirs": true,

    // Extra directories prepended to the PATH environment variable to
    // supplement searching for various Haskell executables, such as ghc and
    // stack.
    //
    // Note: You can use environment variables here, e.g., $HOME and %APPDATA%.
    // Only "%something%" will work on Windows platforms. You can also use the
    // tilde ("~") as an abbreviation for your home directory.
    //
    // If used in combination with "add_standard_dirs", PATH is:
    //
    //   [add_to_PATH directories] + [standard_dirs] + [PATH]
    "add_to_PATH": [],


    // Add word completions to the completions list. This consists of any word
    // in the current document that is longer than 3 characters.
    "add_word_completions": false,

    // Add default completions from the Haskell.sublime-completions file. You can
    // probably leave this setting false, unless you're using the "none" backend.
    "add_default_completions": false,

    // Auto autocompletion popup on '.' in import list
    "auto_completion_popup": false,

    // Additional ghc options for use in ghc-mod and ModuleInspector, for example "-package ghc"
    "ghc_opts": [],

    // ~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=
    // Backend selection:
    // ~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=

    // The backend dictionary, whose structure is:
    //
    //  "name of backend": {
    //    "default": true|false,             // Only one of your backends should be true. SublimeHaskell will use the first
    //                                       // default backend it finds if there is more than one default backend.
    //                                       // (remember: this is a dictionary, so it's not necessarily the one you think
    //                                       // it should be.)
    //    "backend": "hsdev|ghc-mod|none",   // Self explanatory
    //    "options": {
    //      // see below
    //    }
    //  }
    //
    // There are three backends: "hsdev", "ghc-mod" and "none".
    //
    // Options for hsdev:
    //
    //     port:        [default: "4567"]      TCP port where SublimeHaskell will connect to hsdev
    //     host:        [defualt: "localhost"] Host name SublimeHaskell will connect to hsdev
    //     local:       [default: true]        Should SublimeHaskell create a local hsdev server process? Specify false if you
    //                                         want to connect to a remote hsdev server
    //     exec-with:   [default: ""]          Should SublimeHaskell use "stack" or "cabal" to when starting
    //                                         the hsdev executable?
    //     install-dir: [default ""]           Used in conjunction with "exec-with"; this is the directory where SublimeHaskell
    //                                         can find  the hsdev executable.
    //
    // There are no additional options for either the ghc-mod or none backends

    "backends": {
        "Local hsdev": {
            "default": true,
            "backend": "hsdev",
            "options": {
              "local": true,
              "port": 4567,
              "install-dir": "/home/merf/.local/bin"
          }
        },
        "ghc-mod": {
            "default": false,
            "backend": "ghc-mod"
        },
        "Remote hsdev@localhost:4567": {
            "default": false,
            "backend": "hsdev",
            "options": {
                "local": false,
                "host": "localhost",
                "port": 4567
            }
        },
        "Cabal-sandboxed local hsdev": {
            "default": false,
            "backend": "hsdev",
            "options": {
              "local": true,
              "port": 4567,
              "exec-with": "cabal"
              // ^^^^^^^^^^^^^^^^^ Remember to insert a comma (",") after this line once you set install-dir
              // to hsdev's installation directory.
              //
              // Force a diagnostic message because this has to be set in conjunction with exec-with:
              // "install-dir": "$HOME/totally/invalid/directory"
            }
        },
        "Stack-exec local hsdev": {
            "default": false,
            "backend": "hsdev",
            "options": {
              "local": true,
              "port": 4567,
              "exec-with": "stack",
              // ^^^^^^^^^^^^^^^^^ Remember to insert a comma (",") after this line once you set install-dir
              // to hsdev's installation directory.
              //
              // Force a diagnostic message because this has to be set in conjunction with exec-with:
              // "install-dir": "$HOME/totally/invalid/directory"
              "install-dir": "/home/merf/.local/bin"
            }
        },
        "Cabal new-build local hsdev": {
            "default": false,
            "backend": "hsdev",
            "options": {
              "local": true,
              "port": 4567,
              "exec-with": "cabal-new-build"
              // ^^^^^^^^^^^^^^^^^ Remember to insert a comma (",") after this line once you set install-dir
              // to hsdev's installation directory.
              //
              // Force a diagnostic message because this has to be set in conjunction with exec-with:
              // "install-dir": "$HOME/totally/invalid/directory"
            }
        },
        "No backend": {
            "default": false,
            "backend": "none"
        }
    },

    // ~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=
    // User interaction settings:
    // ~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=

    // lint & check on the fly
    "lint_check_fly": true,

    // idle time in seconds before lint & check on the fly
    "lint_check_fly_idle": 5,

    // Use unicode symbols info 'symbol info panel'
    "unicode_symbol_info": true,

    // Enable hdocs, disabled by default since 'hdocs' is not on hackage yet
    // It can be manually installed from http://github.com/mvoidex/hdocs
    "enable_hdocs": false,

    // Enable inpecting modules with SublimeHaskell's ModuleInspector.
    // Needed for autocompletion.
    // This can take a lot of time in really big Haskell projects.
    // Changing this requires a Sublime restart.
    "inspect_modules": true,

    // Additional ghc options for repl
    "ghci_opts": [],

    // Run a code indenter/prettyfier after save.
    "prettify_on_save": false,
    // The prettifier to use when indenting/prettifying on save. Valid options are "stylish-haskell" or "hindent"
    "prettify_executable": "stylish-haskell",
    // Option list for hindent
    //
    // NOTE: This is not a string. Having Python parse an option string will be suboptimal. Use a list of options
    // instead, e.g., ['--foo', '--bar', '--baz']
    "hindent_options": [],
    // Option list for stylish-haskell
    //
    // NOTE: This is not a string. Having Python parse an option string will be suboptimal. Use a list of options
    // instead, e.g., ['--foo', '--bar', '--baz']
    "stylish_options": [],

    // ~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=
    // Logging verbosity
    // ~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=~=

    // Log level, 0 — no log, 1 — errors, 2 — warnings, 3 — info messages, 4 — debug, 5 — trace
    "log": 5,

    // hsdev logging output:
    //
    // !!!PAY ATTENTION TO YOUR HSDEV VERSION!!!
    //
    // hsdev version 0.2.2.0 and below: The "hsdev_log_config" preference controls hsdev's logging. It has several interesting
    // values: 'use default' (somewhat verbose, informational), 'use silent' (quiet)
    //
    // hsdev version 0.2.3.0 and higher: The "hsdev_log_level" preference controls hsdev's logging. Valid values are
    // 'trace', 'debug', 'info', 'warning', 'error' and 'fatal'.
    "hsdev_log_config": "use silent",
    "hsdev_log_level": "warning",

    // Backend debugging flags/tokens: This is a list of tokens used for backend debugging, primarily in the hsdev
    // backend.
    //
    // "all_messages": Emit all messages (send and receive)
    // "send_messages": Emit all sent messages
    // "recv_messages": Emit all received messages
    // "socket_pool": Keep track of the socket pool avail/used sizes.
    // "callbacks": Show interesting information about 'hsdev' callbacks
    // "event_viewer": Event handling dignostics and debug
    // "completions": Completion generator diagnostics and debug
    "component_debug": [""]

    // SublimeHaskell per-project settings (or "things you can put in your .sublime-project file")
    //
    // - active_stack_config
    // - auto_build_mode
    // - haskell_build_tool
    //
    // Example:
    // {
    //   "folders":
    //   [
    //     {
    //       "path": "."
    //     }
    //   ],
    //   "SublimeHaskell": {
    //     // SublimeHaskell project settings here:
    //     "haskell_build_tool": "cabal-new-build"
    //   }
    // }
}
```

User settings:
```
{
	"add_to_PATH": ["/home/merf/.local/bin"]
}
```

Console log:
```
DPI scale: 2
startup, version: 3143 linux x64 channel: stable
executable: /opt/sublime_text/sublime_text
working dir: /
packages path: /home/merf/.config/sublime-text-3/Packages
state path: /home/merf/.config/sublime-text-3/Local
zip path: /opt/sublime_text/Packages
zip path: /home/merf/.config/sublime-text-3/Installed Packages
ignored_packages: ["Vintage"]
pre session restore time: 0.048848
startup time: 0.119474
first paint time: 0.123299
reloading plugin Default.auto_indent_tag
reloading plugin Default.block
reloading plugin Default.comment
reloading plugin Default.convert_syntax
reloading plugin Default.copy_path
reloading plugin Default.delete_word
reloading plugin Default.detect_indentation
reloading plugin Default.duplicate_line
reloading plugin Default.echo
reloading plugin Default.exec
reloading plugin Default.fold
reloading plugin Default.font
reloading plugin Default.goto_line
reloading plugin Default.history_list
reloading plugin Default.indentation
reloading plugin Default.install_package_control
reloading plugin Default.kill_ring
reloading plugin Default.mark
reloading plugin Default.new_templates
reloading plugin Default.open_context_url
reloading plugin Default.open_in_browser
reloading plugin Default.pane
reloading plugin Default.paragraph
reloading plugin Default.paste_from_history
reloading plugin Default.profile
reloading plugin Default.quick_panel
reloading plugin Default.run_syntax_tests
reloading plugin Default.save_on_focus_lost
reloading plugin Default.scroll
reloading plugin Default.set_unsaved_view_name
reloading plugin Default.settings
reloading plugin Default.show_scope_name
reloading plugin Default.side_bar
reloading plugin Default.sort
reloading plugin Default.swap_line
reloading plugin Default.switch_file
reloading plugin Default.symbol
reloading plugin Default.transform
reloading plugin Default.transpose
reloading plugin Default.trim_trailing_white_space
reloading plugin Default.ui
reloading plugin CSS.css_completions
reloading plugin Diff.diff
reloading plugin HTML.encode_html_entities
reloading plugin HTML.html_completions
reloading plugin 0_package_control_loader.00-package_control
reloading plugin Package Control.1_reloader
reloading plugin Package Control.2_bootstrap
reloading plugin Package Control.Package Control
reloading plugin SublimeHaskell.__init__
reloading plugin SublimeHaskell.add_import
reloading plugin SublimeHaskell.autocomplete
reloading plugin SublimeHaskell.backend_cmds
reloading plugin SublimeHaskell.build
reloading plugin SublimeHaskell.cabal
reloading plugin SublimeHaskell.check_lint
reloading plugin SublimeHaskell.cmdwin_types
reloading plugin SublimeHaskell.commands
reloading plugin SublimeHaskell.indenters
reloading plugin SublimeHaskell.info_popup
reloading plugin SublimeHaskell.parseoutput
reloading plugin SublimeHaskell.repl
reloading plugin SublimeHaskell.sublime_haskell_common
reloading plugin SublimeHaskell.symbols
reloading plugin SublimeHaskell.types
reloading plugin SublimeHaskell.zzplugin
plugins loaded
Sublime Haskell: std_places = ['/home/merf/.local/bin', '/usr/local/bin']
Sublime Haskell: add_to_PATH = ['/home/merf/.local/bin']
Sublime Haskell: ProcHelper.exec_with_wrapper: ['/home/merf/.local/bin/hsdev', 'version'] in None
Sublime Haskell: hsdev version: 0.2.5.1
Traceback (most recent call last):
  File "/opt/sublime_text/sublime_plugin.py", line 210, in on_api_ready
    m.plugin_loaded()
  File "/home/merf/.config/sublime-text-3/Packages/SublimeHaskell/zzplugin.py", line 34, in plugin_loaded
    backend_mgr.get_backends()
  File "/home/merf/.config/sublime-text-3/Packages/SublimeHaskell/internals/backend_mgr.py", line 140, in get_backends
    self.possible_backends = self.filter_possible(Settings.PLUGIN.backends)
  File "/home/merf/.config/sublime-text-3/Packages/SublimeHaskell/internals/backend_mgr.py", line 208, in filter_possible
    if ubclazz is not None and ubclazz.is_available(**ubackend.get('options', {})):
  File "/home/merf/.config/sublime-text-3/Packages/SublimeHaskell/hsdev/backend.py", line 93, in is_available
    hsdev_ver = HsDevBackend.hsdev_version(kwargs.get('exec-with'), kwargs.get('install-dir'))
  File "/home/merf/.config/sublime-text-3/Packages/SublimeHaskell/hsdev/backend.py", line 212, in hsdev_version
    hsdev_proc = ProcHelper.exec_with_wrapper(exec_with, install_dir, ['hsdev', 'version'])
  File "/home/merf/.config/sublime-text-3/Packages/SublimeHaskell/internals/proc_helper.py", line 175, in exec_with_wrapper
    raise RuntimeError('ProcHelper.exec_with_wrapper: invalid install_dir (None)')
RuntimeError: ProcHelper.exec_with_wrapper: invalid install_dir (None)
Package Control: Skipping automatic upgrade, last run at 2017-12-11 22:25:28, next run at 2017-12-11 23:25:28 or after
```

**Comment by bscottm** [comment]
@merflogin, @paxter: Yup, I found the problem. I needed to comment out all of my user preferences. It turns out that the "Cabal new-build local hsdev" was the culprit:

        "Cabal new-build local hsdev": {
            "default": false,
            "backend": "hsdev",
            "options": {
              "local": true,
              "port": 4567,
              "exec-with": "cabal-new-build"
              // ^^^^^^^^^^^^^^^^^ Remember to insert a comma (",") after this line once you set install-dir
              // to hsdev's installation directory.
              //
              // Force a diagnostic message because this has to be set in conjunction with exec-with:
              // "install-dir": "$HOME/totally/invalid/directory"
            }
        },

The extra '"local": true' line was the problem: Yes, a stack or cabal-wrapped backend is local, but the code wasn't using the configuration options in a way that would allow all three ("local", "exec-with" and "install-dir") to work together correctly. Testing now, will push new version in a wee bit.

You can take out all of your user preference mods. Just delete the '"local": true' in the "Cabal new-build local hsdev" and you should be on your way.

**Comment by paxter** [comment]
Deleted the local: 'true', but it's the same as before.

**Comment by bscottm** [comment]
@paxter: Try again with the 2.1.6/2.1.7 updates. Sorry to put you on the update hamster wheel. We'll get this fixed. Promise.

**Comment by paxter** [comment]
Looks like the Traceback errors are gone, but it's not working properly. The menu entry is still broken and nothing works. I edited the config as above.

**Comment by bscottm** [comment]
@paxter: I'm not entirely clear about "not working properly" and confused with respect to which menu entry is broken -- it could be the _SublimeREPL_ menu, or what should be the _SublimeREPL_ menu if _SublimeREPL_ is installed (that's just my guess.)

What specifically isn't working for you? If you could set `log: 4` in your _SublimeHaskell_ user preferences, then copy and paste it into this issue, that would give me a lot more insight into diagnosing the problem.

**Comment by bscottm** [comment]
@merflogin: Everything working?

**Comment by paxter** [comment]
As I posted in the OSX issue, the menu entry is broken: https://i.imgur.com/ZYnLerk.png

I set log to 4, but looks like there is no error message. Maybe it's an OSX related problem.

`startup, version: 3134 osx x64 channel: dev
executable: /Applications/Sublime Text.app/Contents/MacOS/Sublime Text
working dir: /
packages path: /Users/empire/Library/Application Support/Sublime Text 3/Packages
state path: /Users/empire/Library/Application Support/Sublime Text 3/Local
zip path: /Applications/Sublime Text.app/Contents/MacOS/Packages
zip path: /Users/empire/Library/Application Support/Sublime Text 3/Installed Packages
ignored_packages: ["Vintage"]
pre session restore time: 0.158076
startup time: 0.191861
reloading plugin Default.auto_indent_tag
reloading plugin Default.block
reloading plugin Default.comment
reloading plugin Default.convert_syntax
reloading plugin Default.copy_path
reloading plugin Default.delete_word
reloading plugin Default.detect_indentation
reloading plugin Default.duplicate_line
reloading plugin Default.echo
reloading plugin Default.exec
reloading plugin Default.fold
reloading plugin Default.font
reloading plugin Default.goto_line
reloading plugin Default.history_list
reloading plugin Default.indentation
reloading plugin Default.install_package_control
reloading plugin Default.kill_ring
reloading plugin Default.mark
reloading plugin Default.new_templates
reloading plugin Default.open_context_url
reloading plugin Default.open_in_browser
reloading plugin Default.pane
reloading plugin Default.paragraph
reloading plugin Default.paste_from_history
reloading plugin Default.profile
reloading plugin Default.quick_panel
reloading plugin Default.run_syntax_tests
reloading plugin Default.save_on_focus_lost
reloading plugin Default.scroll
reloading plugin Default.set_unsaved_view_name
reloading plugin Default.settings
reloading plugin Default.show_scope_name
reloading plugin Default.side_bar
reloading plugin Default.sort
reloading plugin Default.swap_line
reloading plugin Default.switch_file
reloading plugin Default.symbol
reloading plugin Default.transform
reloading plugin Default.transpose
reloading plugin Default.trim_trailing_white_space
reloading plugin Default.ui
reloading plugin CSS.css_completions
reloading plugin Diff.diff
reloading plugin HTML.encode_html_entities
reloading plugin HTML.html_completions
reloading plugin 0_package_control_loader.00-package_control
reloading plugin Package Control.1_reloader
reloading plugin Package Control.2_bootstrap
reloading plugin Package Control.Package Control
reloading plugin SublimeHaskell.__init__
reloading plugin SublimeHaskell.add_import
reloading plugin SublimeHaskell.autocomplete
reloading plugin SublimeHaskell.backend_cmds
reloading plugin SublimeHaskell.build
reloading plugin SublimeHaskell.cabal
reloading plugin SublimeHaskell.check_lint
reloading plugin SublimeHaskell.cmdwin_types
reloading plugin SublimeHaskell.commands
reloading plugin SublimeHaskell.indenters
reloading plugin SublimeHaskell.info_popup
reloading plugin SublimeHaskell.parseoutput
reloading plugin SublimeHaskell.repl
reloading plugin SublimeHaskell.sublime_haskell_common
reloading plugin SublimeHaskell.symbols
reloading plugin SublimeHaskell.types
reloading plugin SublimeHaskell.zzplugin
plugins loaded
Sublime Haskell: std_places = ['/Users/empire/Library/Haskell/bin', '/Users/empire/Library/Haskell/bin', '/Library/Haskell/bin']
Sublime Haskell: add_to_PATH = ['/Users/empire/Library/Haskell/bin']
Sublime Haskell: ProcHelper.exec_with_wrapper: ['/Users/empire/Library/Haskell/bin/hsdev', 'version'] in None
Sublime Haskell: hsdev version: 0.2.5.1
Sublime Haskell: ProcHelper.exec_with_wrapper: ['cabal', 'exec', 'hsdev', '--', 'version'] in /Users/empire/Library/Haskell/bin
Sublime Haskell: hsdev version: 0.2.5.1
plugin 'backends' ['Stack-exec local hsdev', 'No backend', 'ghc-mod', 'Cabal-sandboxed local hsdev', 'Remote hsdev@localhost:4567', 'Local hsdev', 'Cabal new-build local hsdev']
Possible/usable 'backends': ['Remote hsdev@localhost:4567', 'No backend', 'Local hsdev', 'Cabal new-build local hsdev']
environment variables loaded using: /bin/bash -l
Package Control: Skipping automatic upgrade, last run at 2017-12-13 13:00:27, next run at 2017-12-13 14:00:27 or after
`

**Comment by bscottm** [comment]
@paxter: 

(a) The two _SublimeHaskell_ submenus are attached to _PackageControl_ and _SublimeREPL_. My guess is that the _SublimeREPL_ submenus are missing or empty, based on the picture. _PackageControl_ submenus would show up if you were navigating the OS X _Preferences_ menu item under *Sublime Text*. Check to see if you have _SublimeREPL_ installed:

    <Shift>-<Command>-<P> Package Control: List Packages

_SublimeREPL_  should be in that list. If not, try installing it. That should fix the submenu problem.

(b) Which set of backend options do you really want to use? "Local hsdev" with `hsdev` installed in `~/Library/Haskell/bin`? The one line:

    Sublime Haskell: ProcHelper.exec_with_wrapper: ['cabal', 'exec', 'hsdev', '--', 'version'] in /Users/user/Library/Haskell/bin

indicates that you are trying to execute a `cabal`-wrapped `hsdev` installed there. Not really a problem, just excess overhead to get a `hsdev` backend to run.

(c) Your log doesn't show me which `hsdev` was actually started. Try putting this in your _SublimeHaskell_ user settings (or modify your `backend` _SublimeHaskell_ user settings):

    "backends": {
        "Local hsdev": {
            "default": true,
            "backend": "hsdev",
        }
    }

`~/Library/Haskell/bin` is automagically added to your `PATH` if you're running on OS X (see the `std_places` line in the log -- _SublimeHaskell_ augments the `PATH` it uses with a few extra well known _Haskell_ paths. You shouldn't have to tweak the `add_to_PATH` setting.)

(d) Open a `.hs` file. _SublimeHaskell_ will not start the backend until you switch to a _Haskell_ source file (or cabal file.) You should see a few status messages starting with "Inspecting..." if `hsdev` has started up correctly. Once the source inspection process is done, try:

    <Shift>-<Command>-<P> SublimeHaskell: Check & Lint

Hope this helps you move a little farther forward.

**Comment by merflogin** [comment]
@bscottm, on version 2.1.7 everything works right after ST restart.
Can't really check your point on _Sublime REPL_. I used to have it installed and not setuped, cause it looked like some tweaks are needed to use it behind stack. Maybe that menu in (Tools -> <empty> -> SublimeHaskell) supposed to hold REPL functionality. Doesn't affect me, I'm happy.
 TY!
(a) There is only Package Control and SublimeHaskell
(b/c) It's using 'Local hsdev' as the only one with `'default': True` and 'Possible/usable'
```
Sublime Haskell: std_places = ['/home/merf/.local/bin', '/usr/local/bin']
Sublime Haskell: add_to_PATH = ['/home/merf/.local/bin']
Sublime Haskell: ProcHelper.exec_with_wrapper: ['/home/merf/.local/bin/hsdev', 'version'] in None
Sublime Haskell: hsdev version: 0.2.5.1
Sublime Haskell: ProcHelper.exec_with_wrapper: ['/home/merf/.local/bin/hsdev', 'version'] in None
Sublime Haskell: hsdev version: 0.2.5.1
plugin 'backends' ['No backend', 'Cabal new-build local hsdev', 'Stack-exec local hsdev', 'Cabal-sandboxed local hsdev', 'Remote hsdev@localhost:4567', 'Local hsdev', 'ghc-mod']
Possible/usable 'backends': ['No backend', 'Cabal new-build local hsdev', 'Local hsdev', 'Remote hsdev@localhost:4567']
Package Control: Skipping automatic upgrade, last run at 2017-12-14 00:45:55, next run at 2017-12-14 01:45:55 or after
Sublime Haskell: Starting backend 'Local hsdev'
Sublime Haskell: HsDevBackend.__init__(localhost, 4567)
Sublime Haskell: ProcHelper.exec_with_wrapper: ['/home/merf/.local/bin/hsdev', 'version'] in None
Sublime Haskell: Starting local 'hsdev' server
Sublime Haskell: ProcHelper.exec_with_wrapper: ['/home/merf/.local/bin/hsdev', 'run', '--port', '4567', '--cache', '/home/merf/.config/sublime-text-3/Cache/SublimeHaskell/hsdev', '--log', '/home/merf/.config/sublime-text-3/Cache/SublimeHaskell/hsdev/hsdev.log', '--log-level', 'warning'] in None
Sublime Haskell: Local 'hsdev' server started successfully.
Sublime Haskell: Connecting to 'hsdev' server at localhost:4567
Sublime Haskell: [pool 0]: connecting to hsdev server (attempt 1)...
Sublime Haskell: [pool 1]: connecting to hsdev server (attempt 1)...
Sublime Haskell: [pool 2]: connecting to hsdev server (attempt 1)...
Sublime Haskell: [pool 3]: connecting to hsdev server (attempt 1)...
Sublime Haskell: Connections established to 'hsdev' server.
```
(d) I see inspector/check/hlint correctly

**Comment by bscottm** [comment]
@merflogin: _SublimeREPL_ is optional IIRC, 2.1.4 was when I made _SublimeHaskell_ REPL support both `cabal` and `stack`. Better REPL support was in a recent release. Won't break anything if you don't have it installed.

You can also uncustomize `add_to_PATH` -- `~/.local/bin` is one of the standard places that _SublimeHaskell_ adds to `PATH`.

Looks like things are working for you. I can sleep better.

**Comment by paxter** [comment]
Thanks a lot for your great support bscottm. I installed the missing SublimeREPL and upgraded SublimeHaskell to 2.1.8. After restarting it's working now!
Thanks again for your time and your patience to help me fix my issues. :)

**Comment by bscottm** [comment]
@paxter: No problem, go and do good Haskell!
